Pixi instead of conda

wwood opened the pull request:
A quite messy move towards using pixi rather than conda. The migration is not yet complete, as only some (mainly assembly and binning) have been ported. Along the way, some other things were changed e.g. vamb was updated - was easier than fixing the previous version.

This is only a draft PR. It would be good to address the issues below and merge this before the next release, but I wanted to put it out there so new features (nanoMDBG, CompleteBin) can be started from this fork, not the dev branch to prevent busywork.

Pixi has a few advantags:

1. Environments are solved by us, not by the user. These are cached in pixi.lock
2. Because of this, we can set permissive versions in pixi.toml, only locking them down at release time. pip and conda env file for bioconda are generated automatically from the pixi.toml file
3. Faster overall

Batch mode has been deprecated in this pr - the tests were failing. There's a single commit where this happens:

commit 3b021dadd7358d195fd17e1b9b36c3a01cf2ca32 (HEAD -> pixi, origin/pixi)
Author: Ben Woodcroft <benjwoodcroft@gmail.com>
Date:   Mon May 19 09:24:19 2025 +1000

    batch: Remove it.
    
    Batch had at least 2 problems:
    1. The test in test_integration was failing (need to add --cluster?)
    2. There was a hard-coded ~/git/aviary path in the code.


Notes / improvements

* Moved bbmap to its own environment, now not in main
* vamb updated (but not taxvamb yet, but that is easy)
* In binning and assembly there's lines like this. But they don't do anything (and never did?) I don't think they affect shell: - removed them.
      sys.path.append(os.path.join(os.path.dirname(os.path.abspath(workflow.snakefile)),"../../scripts"))
    sys.path.append(os.path.join(os.path.dirname(os.path.abspath(workflow.snakefile)),"scripts"))
* --build-only is hard to implement now, so instead I was thinking just pixi build -a, build everything. Advanced users can choose their own environments to build if needed.
* DBs now symlinked from db/ so that they can be specific to our servers. This replaces setting conda envs in .activate scripts in the aviary conda env (there is now no longer a need for an aviary conda env)
* For README "Option 2: Install from pip" - why is the environment needed? This seems a bit pointless and redundant with installing via bioconda. Shall we just say "you can install with pip, but there are other dependencies need. These are documented in `environment.yml`"?
* Why is rosella_refine using checkm1 and not checkm2?


Todo:

@AroneyS 
- [x] developer-specific instructions - test strategy/specifics
- [x] ensure it works when running from outside the aviary dev dir
- [x] add symlinks to toml, lock, .pixi for dev?
- [x] update taxvam
- [x] In get_binned_reads.py it uses mfqe, but the rule `get_read_pools` in assembly.smk no environment is set - shouldn't it be?
- [x] same with assemble_pools ?
- [x] implement --build-only @AroneyS 
- [x] singlem version was bumped to 0.18.3. There's a new long read version coming soon which should be intergrated. Waiting for upstream pixi fix
- [x] I'm confused by the reference-filter, in assembly.smk. i think the whole thing is a bit suspicious and requires targeted testing (or dropping the feature) - rename to host-filtering (and update method? new deacon?)
- [x]  "One or more reference filter files to aid in the assembly" in qc_group help is a little confusing, could be clearer in what they are used for.
- [x] 3 soil, 3 ocean, 3 gut applied tests to see if binning gets better between releases
- [x] account for run-post-link-scripts in build
- [x] https://github.com/rhysnewell/aviary/issues/281 - fail gracefully when there are no bins found
- [x] Add separate logging for each attempt + custom error handling to print the failing log
- [x] Why does gtdbtk skip the ani_screen in annotation.smk ? gtdbtk should use mash_db after download of the DB, in a centralised place (maybe add `.msh` to the end of the gtdbtk db? - instead just update to v2.5.1

@wwood 
- [x] Remove stray conda env yaml files that are not longer needed
- [x] set --frozen when running rules
- [x] make environments more permissive
- [x] make sure setup.py and pip install works
- [x] when using bash -c also use -e and -o pipefail @wwood 
- [x] It would be good harmonise `aviary download` and `configure` to be in sync with the new `db/` symlinking method.  Is configure worth keeping around? Or just ask the user to set their own env vars? #278 ?
- [x] --tmpdir None being passed into the script and used? e.g. in prepare_binning_files
- [x] --tmpdir if set to a relative path fails?
- [x] All rules should have proper log files
- [x] change the set_env_vars to be like singlem, so it can be run in arbitrary directories
- [x] check for read permissions on all input files, otherwise croak.
- [x] https://github.com/rhysnewell/aviary/issues/211 - filter contig to large enough before mapping. See filter_contigs_by_size rule in binning.smk
- [x] GTDB-Tk should use split mode to reduce RAM requirements

more

 - [x] webpage rule - maintained or ditch? Doesn't seem to work with `GTDBTK_DATA_PATH=. CHECKM2DB=. EGGNOG_DATA_DIR=. METABULI_DB_PATH=. SINGLEM_METAPACKAGE_PATH=. aviary recover --assembly /home/ben/git/aviary/test/data/assembly.fasta -1 /home/ben/git/aviary/test/data/wgsim.1.fq.gz -2 /home/ben/git/aviary/test/data/wgsim.2.fq.gz --output /tmp/tmpiqwj__v4/test --dryrun --tmpdir /tmp/tmpiqwj__v4 -w www/index.html`
 - [ ] Ditch isolate mode?
 - [x] benchmarking now broken? Should really be run outside aviary anyway?
 - [x] Test strain analysis
 - [x] Test virsorter
 - [ ] config[max_memory] - remove for GTDBtk? Copy bin chicken?
 - [x] Update documentation

wwood committed:
pixi: First steps.

wwood committed:
pixi: First workflow w/o conda.

wwood committed:
admin: Towards pixi.

wwood committed:
test: Use pytest and adjust actions.

wwood committed:
pixi: Further steps.

wwood committed:
pixi: First version with test_integration passing except batch.

wwood committed:
batch: Remove it.

Batch had at least 2 problems:
1. The test in test_integration was failing (need to add --cluster?)
2. There was a hard-coded ~/git/aviary path in the code.

wwood committed:
Merge remote-tracking branch 'origin/dev' into pixi

wwood committed:
taxvamb: Update to v5.

wwood committed:
pixi: Run frozen and with a defined manifest.

AroneyS committed:
test pixi files within aviary package dir

AroneyS committed:
fix manifest path

AroneyS committed:
move pixi properly into aviary package dir
use `pixi install -a --manifest-path aviary/pixi.toml` to install to aviary/.pixi/envs

AroneyS committed:
add manifest path arg to tests

AroneyS committed:
fix coassembly bug

AroneyS commented:
Bin Chicken installs Aviary using `pixi_run.replace("run", "install -a", 1)` and then its sub-envs with
```
aviary_dirname = extern.run(f"{pixi_run} -e aviary python -c 'import aviary, os; print(os.path.dirname(aviary.__file__))'")
extern.run(f"{pixi_run} -e aviary pixi install -a --manifest-path {aviary_dirname}/pixi.toml")
```

This then installs Aviary in `.pixi/envs/aviary` and the sub-envs in `.pixi/envs/aviary/lib/python3.11/site-packages/aviary/.pixi/envs/`. Nice and neat.

Instead of `--build-only` arg, I would suggest a separate subcommand that just runs `pixi_run.replace("run", "install -a", 1)`. Then you could have an arg for advanced users that specifies the envs to install.

wwood commented:
So you suggest a new subcommand (say `install` ?) which runs build -a, and then raise a NotImplementedException for `--build-only` ?

I'm thinking it might be better from a UI perspective to just have `--build-only` run install -a like you suggest? Then the breakage is that too many environments are installed (i.e. not just the ones required to run that workflow), rather than a complete fail with `--build-only`.

Or could do both i.e. `--build-only` effectively runs `install` for now?

AroneyS committed:
remove references to CONDA_ENV_PATH env variable

AroneyS committed:
convert `--build` arg to pixi install

AroneyS committed:
add build subcommand

AroneyS committed:
remove more conda-prefix references

AroneyS committed:
remove more references to conda

AroneyS committed:
fix test

AroneyS committed:
mark build tests as expensive

AroneyS commented:
Any idea why `fastq = "data/short_reads.fastq.gz"` is an input to assemble_short_reads? It isn't used

AroneyS committed:
ignore "None" tmp_dir arg

AroneyS committed:
move resources before --configfile to prevent target_rule being considered under resources
fixes error `not enough values to unpack (expected 2, got 1)`

AroneyS committed:
create done file when skipping refinery

AroneyS committed:
add metabuli db path arg to all commands

AroneyS committed:
move package installation to task to prevent dependency lockup
updating requirements.txt depends on the installed pixi env which depends on requirements.txt to install the local package...
pixi.toml pypi dependencies do not currently support `--no-deps`

AroneyS committed:
fix binner gpu environments
ensure that they are actually using the gpu
installed on machine with H100s and worked on machine with A100s

AroneyS committed:
download all if none specified

AroneyS committed:
fix default download arg

AroneyS committed:
move annotation rules to pixi

AroneyS committed:
adapt get_read_pools to pixi
add final-assembly pixi env

AroneyS committed:
move clustering rules to pixi

AroneyS committed:
add pixi_run import to clustering

wwood committed:
Merge pull request #278 from rhysnewell/download-all-if-none-specified

Download all databases if none specified

wwood committed:
admin/check_versions.py: New script.

wwood commented:
Made a quick tool to test whether versions are up to date and added script to admin dir. Some work to do, some pretty easy.
https://gist.github.com/wwood/34e2ae1d2daf1e88badb7877e0a42bea

Needs prioritisation though - maybe if the name of the environment is the name of the tool and that is outdated it should fail. Should run on a cronjob I guess like every month or something. Ideally runs the tests too after updating the toml optimistically. It's all down to the test suite.

But anyway would be good to sit down and discuss this PR @AroneyS and how to finish. I think the TODO list might be a bit dated.

wwood committed:
binning: Only bin large contigs. Fixes #211.

Reported by: @dylancronin

wwood committed:
binning: Remove wrong touch() on inputs.

wwood committed:
conda: Remove many leftoever yaml files.

wwood committed:
annotation: Remove conda-specific gtdbtk download code.

wwood committed:
gtdbtk: Set DB to be R226.

wwood committed:
admin: set_env_vars.sh: Adjust for moving pixi to aviary/

wwood committed:
test: Add run_tests_at_cmr.sh

AroneyS committed:
provide GiB instead of GB to metabuli

wwood committed:
.gitignore: Adjust for run_cmr_tests

wwood committed:
checkm2: Fix pixi invocation.

wwood committed:
pixi: Loosen many dependencies in many environments.

wwood committed:
modules: Replace bash -c with -e -o pipefail.

wwood committed:
assembly: Avoid tmpdir being set to 'None'.

wwood committed:
gtdbtk: Require 2.4.1 in line with R226 db.

wwood committed:
gtdbtk: Fix environment.

wwood committed:
test_annotate: New file, passes.

wwood committed:
run_tests_at_cmr: Run each test as a separate qsub.

Does not pass yet.

wwood committed:
aviary: Test for input readset permissions.

AroneyS committed:
add --build-gpu to standard and --gpu to build
otherwise don't install envs ending in `-gpu`

AroneyS committed:
add run test samples script

AroneyS committed:
add symlinks for pixi files/folder

AroneyS committed:
update run-post-link-scripts to insecure within build

AroneyS committed:
update installation docs to match readme

AroneyS committed:
fix test samples commands

AroneyS committed:
rename reference-filter to host-filter

AroneyS committed:
integrate qc properly with assembly
previously qc'd reads were ignored if host-filter was not used

AroneyS committed:
update --host-filter help

AroneyS committed:
fix host-filter integration test

wwood committed:
run_tests_at_cmr: Fix test gathering.

AroneyS committed:
add run_samples_at_cmr script

wwood committed:
aviary: Several minor improvements.

wwood committed:
bugfix, more and better tests.

AroneyS committed:
restrict medaka version

AroneyS committed:
use rosella env for rosella_refine rules

AroneyS committed:
fix metabat2 refined bins.tsv file
metabat2 adds \t total_depth=X \t sample_depths=X to contig names
causes dastool error: ERROR: Cannot read scaffold2bin file: data/metabat2_refined_bins.tsv.	Please check file/format. Should be: scaffold_name \t bin_id

AroneyS committed:
account for no bins found by binner with max iterations >0

AroneyS committed:
prevent pip from searching .pixi folder (takes ages)

AroneyS committed:
update sample test runs to use recovery samples

AroneyS committed:
fix assemble_short_reads inputs

AroneyS committed:
create output folders before touching binner done files

AroneyS committed:
create output folders before || touches instead

AroneyS committed:
clearer response to no bins being found
lets checkm_das_tool fail and catches the error, checking for specific wording in das_tool.log

AroneyS committed:
add developer-specific instructions

AroneyS committed:
fix prepare_binning_files_split

AroneyS committed:
fix metabat2 Fasta_to_Scaffolds output having extra columns due to tabs in the contig names
sometimes these tabs are not present, so awk is used

AroneyS committed:
update running of test samples
switched out a sample due to dodgy sequences

AroneyS committed:
separate log files for each attempt with custom error handling

AroneyS committed:
set env variables for test_run_workflow_errors

AroneyS committed:
check that intermediate files are populated, not just that they exist

AroneyS committed:
update gtdbtk to v2.5.1

AroneyS committed:
add missing gpu test

AroneyS committed:
update readme

AroneyS committed:
change configure to use conda env config vars

AroneyS committed:
skip processing bashrc lines without export and those with PATH

AroneyS committed:
add env type to test sample running output

AroneyS commented:
Aviary test set results (commit dcfdf7d).
Genomes are >50% quality, score is the sum of the quality for those genomes.
258 total genomes, 19,243 total score.

```
Env	Sample	Score	Genomes
marine	ERR599108	1676.8999999999999	24
marine	ERR12716859	481.82	7
marine	SRR13153254	1479.9000000000003	20
soil	SRR6490520	5000.009999999998	72
soil	ERR2210124	179.64	3
soil	SRR23478535	1360.65	18
human_gut	ERR4283370	4348.920000000001	53
human_gut	SRR13348569	4103.610000000001	55
human_gut	SRR6028613	611.9100000000001	7
```

And with including small contigs.
252 total genomes, 19,174 total score.

```
Env	Sample	Score	Genomes
marine	ERR599108	1542.5399999999997	21
marine	ERR12716859	984.5699999999998	15
marine	SRR13153254	1502.86	20
soil	SRR6490520	4618.339999999999	62
soil	ERR2210124	122.23	2
soil	SRR23478535	1317.33	17
human_gut	ERR4283370	4158.17	51
human_gut	SRR13348569	4213.19	56
human_gut	SRR6028613	716.35	8
```

So small contigs filtering change doesn't seem to alter the results much. Sometimes better, sometimes worse recovery.

AroneyS committed:
Merge pull request #284 from rhysnewell/update-gtdbtk-2.5.0

Update gtdbtk to 2.5.0

wwood committed:
refine_das_tool: Use checkm2 not checkm1.

Reported by @michoug.

Attempts to address https://github.com/AroneyS/binchicken/issues/193

wwood committed:
Fix regression, CHECKM2DB specified down to uniref100.KO.1.dmnd in line with CheckM2 itself.

wwood commented:
Guess we just merge now? You are using this on bin chicken and run tests at CMR works right?

AroneyS commented:
Yes, as long as you are happy with the db stuff

AroneyS commented:
Then if we release, I can try getting aviary-installation working

wwood committed:
pixi.toml: project -> workspace.

wwood commented:
Mind signing off please @AroneyS ?

I just put in a few more commits - rosella now uses checkm2 not 1.

wwood committed:
config: Do not check for new configuration.

This was conda-specific and/or not used any more (I think).

AroneyS commented:
Rerunning tests now

AroneyS committed:
remove reference to deleted function

AroneyS commented:
The CheckM2 change breaks finalise_stats that expects bins/checkm.out to be checkm1 output

AroneyS committed:
don't save to bashrc

AroneyS committed:
remove checkm1

AroneyS committed:
fix tests

AroneyS committed:
move set_env_vars.sh activation to dev env only

AroneyS committed:
try to fix collection errors

AroneyS committed:
add more instructions to run_tests_at_cmr

AroneyS committed:
error das_tool.py properly when no bins found

AroneyS committed:
remove semibin log checking
log paths have moved

AroneyS committed:
add both spades and megahit testing

AroneyS committed:
add spacing for args

AroneyS committed:
allow empty short read arg for polish

AroneyS committed:
run samples at cmr in parallel

AroneyS committed:
bump singlem version to v0.20.3

AroneyS committed:
fix refine with checkm2 script

AroneyS commented:
Updated test samples commit aabe36d. Genomes are >50% quality, score is the sum of the quality for those genomes.
metaspades: 255 total genomes, 19,182 total score (roughly the same as before, some samples slightly up, some slightly down).
megahit: 226 total genomes, 17,046 total score

```bash
Env	Sample	Assembler	Score	Genomes
marine	ERR599108	megahit	1506	22
marine	ERR599108	metaspades	1525	21
marine	ERR12716859	megahit	218	3
marine	ERR12716859	metaspades	763	12
marine	SRR13153254	megahit	1084	15
marine	SRR13153254	metaspades	1567	21
soil	SRR6490520	megahit	4770	64
soil	SRR6490520	metaspades	4910	70
soil	ERR2210124	megahit	63	1
soil	ERR2210124	metaspades	52	1
soil	SRR23478535	megahit	1231	17
soil	SRR23478535	metaspades	1380	18
human_gut	ERR4283370	megahit	3931	48
human_gut	ERR4283370	metaspades	4368	53
human_gut	SRR13348569	megahit	3645	49
human_gut	SRR13348569	metaspades	3837	49
human_gut	SRR6028613	megahit	598	7
human_gut	SRR6028613	metaspades	780	10
```

AroneyS committed:
remove comment

AroneyS committed:
fix test

AroneyS committed:
fix set_env_vars script path

AroneyS committed:
move set_env_vars to admin

AroneyS committed:
continue to next workflow on success

AroneyS committed:
replace reference to aviary.yml to admin/requirements.txt

AroneyS committed:
remove strain diversity and add aviary complete tests

AroneyS committed:
remove viral and webpage modules

AroneyS committed:
remove benchmarking module

AroneyS committed:
remove lorikeet concepts

wwood committed:
prepare_binning_files: Stricter. Update help.

Suggested by: Steven Robbins.

wwood committed:
pixi: Don't use --frozen, add setuptools so postinstall task work.
